### PR TITLE
Archive sooner

### PR DESCRIPTION
**Pull request by justizin**
Still testing this, but one of the concerns I had with the default behavior was waiting to archive the data until it was time to delete.  If the data is already segmented by day, anything older than a day should always be archived.

### CODE REVIEW

File: src/backup-all-indexes.sh [hunk 1]
```diff
@@ -62,9 +62,17 @@ backup_index ()
...
+delete_index ()
+{
+  : ${1:?"Error: expected index name passed as parameter"}
+  local INDEX_NAME=$1
+  local SNAPSHOT_URL=${REPOSITORY_URL}/${INDEX_NAME}
+  local INDEX_URL=${DATABASE_URL}/${INDEX_NAME}
 
   echo "Deleting ${INDEX_NAME} from Elasticsearch."
-  curl -w "\n" -sS -XDELETE ${INDEX_URL}
+  echo curl -w "\n" -sS -XDELETE ${INDEX_URL}
```
Comment (krallin) [inline src/backup-all-indexes.sh]: Is this change intentional? (and what is trying to achieve?)

File: src/backup-all-indexes.sh [hunk 2]
```diff
@@ -62,9 +62,17 @@ backup_index ()
     echo "$(now): Waiting for snapshot to finish..."
     timeout "${WAIT_SECONDS}" bash -c "until grep -q SUCCESS <(curl -sS ${SNAPSHOT_URL}); do sleep 1; done" || return 1
   fi
+}
+
+delete_index ()
+{
+  : ${1:?"Error: expected index name passed as parameter"}
+  local INDEX_NAME=$1
+  local SNAPSHOT_URL=${REPOSITORY_URL}/${INDEX_NAME}
```
Comment (krallin) [inline src/backup-all-indexes.sh]: You can probably skip `SNAPSHOT_URL`:  it never gets called in this function.

File: src/backup-all-indexes.sh [hunk 3]
```diff
@@ -87,9 +95,10 @@ curl -w "\n" -sS -XPUT ${REPOSITORY_URL} -d "{
     \"server_side_encryption\": true
   }
 }"
-
-CUTOFF_DATE=$(date --date="${MAX_DAYS_TO_KEEP} days ago" +"%Y.%m.%d")
-echo "$(now) Archiving all indexes with logs before ${CUTOFF_DATE}."
+ARCHIVE_CUTOFF_DATE=$(date --date="1 days ago" +"%Y.%m.%d") # why would we wait weeks to archive?
```
Comment (krallin) [inline src/backup-all-indexes.sh]: I get where you're coming from but you can probably skip that comment here :)

File: src/backup-all-indexes.sh [hunk 4]
```diff
@@ -100,6 +109,9 @@ for index_name in $(curl -sS ${DATABASE_URL}/_cat/indices | grep logstash- | sed
       else
         echo "$(now): ${index_name} archival failed."
       fi
+      if [[ "${index_name:9}" > "${DELETE_CUTOFF_DATA}" ]]; then
```
Comment (krallin) [inline src/backup-all-indexes.sh]: I'm fairly certain this should be `DELETE_CUTOFF_DATE`, not `DELETE_CUTOFF_DATA` (or this will delete all indices that are archive-able).

TBH, it's a real problem that this script does not run with `set -o nounset`, but for now, that's what it is :/

File: src/backup-all-indexes.sh (hunk 3 above)
Comment (krallin) [reply src/backup-all-indexes.sh]: More importantly: I'm a little concerned about using exactly "1 day" as the cutoff. Using exactly one day sort of assumes that everybody agrees on what the date is — i.e. that they use the same timezone and their time is in sync —, which may not always be 100% accurate.

Unfortunately, [as I noted in this other PR](https://github.com/aptible/elasticsearch-logstash-s3-backup/pull/16/files#r144498117), snapshotting too early in this script means  you may not end up snapshotting everything...

I figure we could either create a final snapshot before deleting an index, _or_ give it 2 days before snapshotting..?

**Review by krallin — CHANGES_REQUESTED** [review]

File: src/backup-all-indexes.sh (hunk 3 above)
Comment (justizin) [reply src/backup-all-indexes.sh]: Yeah, 2 days is good, but I'm unclear how the container running this script and the container the ES data is in would disagree.

My main concern is having ES die, and losing 7, 14, 30, whatever days of data because the archive waits until delete.

Comment (krallin) [reply src/backup-all-indexes.sh]: > Yeah, 2 days is good, but I'm unclear how the container running this script and the container the ES data is in would disagree.

If you're assuming this is running on Enclave, then yes, this is unlikely to happen since we're running Logstash using UTC as the timezone, and you're using this container which also uses UTC as the timezone.

However, that might not be the case for everyone using this: if someone runs Logstash in UTC, but executes this script from a different TZ, they might end up backing up their logs too early because the script thinks today is day N+1 (and thus backs up the logstash-N index), but Logstash thinks today is day N and continues to write to the logstash-N index, thus resulting in any logs written after the backup script runs to never be archived.

> My main concern is having ES die, and losing 7, 14, 30, whatever days of data because the archive waits until delete.

Worth mentioning that, if you're assuming this is running on Enclave, this is largely moot considering the ES disk is actually backed up on a daily basis, though I can understand that you wanting to have your own archives.

File: src/backup-all-indexes.sh (hunk 1 above)
Comment (justizin) [reply src/backup-all-indexes.sh]: nope, this is ... probably why nothing is happening lol. i left debug echoes.

**Comment by rogthefrog** [comment]
FWIW I find the inconsistent nomenclature (`snapshot`, `archive`, `backup`) hard to follow.

File: src/backup-all-indexes.sh (hunk 4 above)
Comment (rogthefrog) [inline src/backup-all-indexes.sh]: Should the deletion be performed in all cases, even if the archival failed?

Comment (justizin) [reply src/backup-all-indexes.sh]: fine point!
